### PR TITLE
Synchronising changes from `js-logger`

### PR DESCRIPTION
**Pull request by emmacasolin**
### Description
Updating CI/CD including:

- Continuous benchmarking
- Staging branch
- Advanced type checking

### Tasks
- [x] 1. Create new staging branch
- [x] 2. Synchronise changes from `js-logger`

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by emmacasolin** [comment]
Should the benchmarks here be changed to be done continuously in the CI/CD (like `js-logger`) or be left as is? @CMCDragonkai

**Comment by CMCDragonkai** [comment]
Yea benches should be migrated over to the same style. It's fairly easy to copy, similar `index.js` and more.

**Comment by CMCDragonkai** [comment]
Should update all the `package.json` dependencies to match.

**Comment by CMCDragonkai** [comment]
I'll run the benchmarks here, so you can move on to other libraires.

File: benches/index.ts [hunk 1]
```diff
@@ -1,18 +1,41 @@
-#!/usr/bin/env node
+#!/usr/bin/env ts-node
 
 import fs from 'fs';
+import path from 'path';
 import si from 'systeminformation';
 import WorkerManagerBench from './WorkerManager';
```
Comment (CMCDragonkai) [inline benches/index.ts:6]: Make this `import workerManager from './worker_manager.ts`.

Comment (CMCDragonkai) [reply benches/index.ts:6]: The new metrics use lower case names for the reports, which I retain as part of the file name.

File: benches/WorkerManager.ts [hunk 2]
```diff
@@ -24,7 +25,7 @@ async function main() {
   // 1 KiB of data is still too small
   const bytes = crypto.randomBytes(1024 * 1024);
   const summary = await b.suite(
-    'WorkerManager',
+    path.basename(__filename, path.extname(__filename)),
     b.add('Call Overhead', async () => {
       // This calls a noop, this will show the overhead costs
       // All parallelised operation can never be faster than this
```
Comment (CMCDragonkai) [inline benches/WorkerManager.ts:31]: You need `utils/utils.ts` and `utils/index.ts`. They contain the common suite. Then you just replace the bottom `cycle` and `complete` and `save` with the `...suiteCommon`.

**Comment by CMCDragonkai** [comment]
I'll need to add this into codesee too after you merge it.

At the same time... try generating the docs as well, and squash it all.

**Comment by CMCDragonkai** [comment]
And the scripts all require executable permission.

**Comment by CMCDragonkai** [comment]
It looks like `ts-node` using `swc` fails to run the benchmarks because the decorator plus symbol properties are not being compiled properly. Disabling and removing swc makes the benchmarks work again. This may be related to https://github.com/swc-project/swc/issues/5189.

So for this project we'll need to switch off the usage of swc. But this creates a problem, any benchmarking for PK which relies on the async-init will end up not working. We will need swc upstream to be fixed.

**Comment by CMCDragonkai** [comment]
I posted a linked bug here https://github.com/TypeStrong/ts-node/issues/1844.

So for now `swc` could affect any benchmarks running in ts-node too... Basically any call ready wrapped functions. Be wary of this, you may wish to disable the `swc: true` in those other repos. But we will keep `swc` where we can until we can make use of it.

**Pull request merged**